File: src/Material/EquationOfState.cc
//---------------------------------Spheral++----------------------------------//
// EquationOfState -- Abstract base class for the equation of state classes.
//
// Created by JMO, Mon Dec  6 21:36:45 PST 1999
//----------------------------------------------------------------------------//

#include "EquationOfState.hh"
#include "Geometry/Dimension.hh"
#include "NodeList/NodeList.hh"
#include "Utilities/bisectRoot.hh"

namespace Spheral {

namespace {

template<typename Dimension>
struct Pfunctor {
  const EquationOfState<Dimension>& mEOS;
  NodeList<Dimension> mNodes;
  mutable Field<Dimension, double> mRho0, mEps, mP;
  double mP0;

  Pfunctor(const EquationOfState<Dimension>& eos, const double rho0, const double P0):
    mEOS(eos),
    mNodes("__dummynodes__", 1, 0),
    mRho0("rho", mNodes, rho0),
    mEps("eps", mNodes),
    mP("P", mNodes),
    mP0(P0) {}

  double operator()(const double eps) const {
    mEps[0] = eps;
    mEOS.setPressure(mP, mRho0, mEps);
    return mP[0] - mP0;
  }
};

}

//------------------------------------------------------------------------------
// Default constructor.
//------------------------------------------------------------------------------
template<typename Dimension>
EquationOfState<Dimension>::EquationOfState(const PhysicalConstants& constants,
                                            const double minimumPressure,
                                            const double maximumPressure,
                                            const MaterialPressureMinType minPressureType,
                                            const double externalPressure):
  mConstants(constants),
  mMinimumPressure(minimumPressure),
  mMaximumPressure(maximumPressure),
  mExternalPressure(externalPressure),
  mMinPressureType(minPressureType) {
  REQUIRE(mMinimumPressure <= mMaximumPressure);
}

//------------------------------------------------------------------------------
// Destructor.
//------------------------------------------------------------------------------
template<typename Dimension>
EquationOfState<Dimension>::~EquationOfState() {
}

//------------------------------------------------------------------------------
// Look up an energy that gives the requested pressure at the specified density.
//------------------------------------------------------------------------------
template<typename Dimension>
typename Dimension::Scalar
EquationOfState<Dimension>::
specificThermalEnergyForPressure(const typename Dimension::Scalar Ptarget,
                                 const typename Dimension::Scalar rho,
                                 const typename Dimension::Scalar epsMin,
                                 const typename Dimension::Scalar epsMax,
                                 const typename Dimension::Scalar epsTol,
                                 const typename Dimension::Scalar Ptol,
                                 const unsigned maxIterations,
                                 const bool verbose) const {
  const Pfunctor<Dimension> pfunc(*this, rho, Ptarget);
  return bisectRoot(pfunc, epsMin, epsMax, epsTol, Ptol, maxIterations, verbose);
}

}

File: src/Pybind11Wraps/Material/EquationOfState.py
#-------------------------------------------------------------------------------
# EquationOfState abstract class
#-------------------------------------------------------------------------------
from PYB11Generator import *
from EOSAbstractMethods import *

@PYB11template("Dimension")
@PYB11module("SpheralMaterial")
class EquationOfState:

    PYB11typedefs = """
    typedef typename %(Dimension)s::Scalar Scalar;
    typedef Field<%(Dimension)s, Scalar> ScalarField;
"""

    #...........................................................................
    # Constructors
    def pyinit(self,
               constants = "const PhysicalConstants&",
               minimumPressure = "const double",
               maximumPressure = "const double",
               minPressureType = "const MaterialPressureMinType",
               externalPressure = ("const double", "0.0")):
        "EOS base constructor"

    #...........................................................................
    # Virtual methods
    @PYB11virtual
    @PYB11const
    def specificThermalEnergyForPressure(self,
                                         Ptarget = "const Scalar",
                                         rho = "const Scalar",
                                         epsMin = "const Scalar",
                                         epsMax = "const Scalar",
                                         epsTol = "const Scalar",
                                         Ptol = "const Scalar",
                                         maxIterations = ("const unsigned", "100")):
        return "Scalar"

    @PYB11virtual
    @PYB11const
    def molecularWeight(self):
        "Optionally provide a molecular weight for an equation of state"
        return "Scalar"

    #...........................................................................
    # Methods
    @PYB11const
    def applyPressureLimits(self, P="const Scalar"):
        "Return the limited pressure"
        return "Scalar"

    #...........................................................................
    # Properties
    constants = PYB11property("const PhysicalConstants&", "constants", doc="The units choice")
    minimumPressure = PYB11property("double", "minimumPressure", "minimumPressure", doc="The minimum allowed pressure")
    maximumPressure = PYB11property("double", "maximumPressure", "maximumPressure", doc="The maximum allowed pressure")
    minimumPressureType = PYB11property("MaterialPressureMinType", "minimumPressureType", "minimumPressureType", doc="The algorithm for enforcing the minimum pressure")
    externalPressure = PYB11property("double", "externalPressure", "externalPressure")
    
#-------------------------------------------------------------------------------
# Add the virtual interface
#-------------------------------------------------------------------------------
PYB11inject(EOSAbstractMethods, EquationOfState, pure_virtual=True)
